### PR TITLE
[FEATURE] Deprecate atomic primitives. Replace with STL atomic.

### PR DESCRIPTION
**Pull request by rrahn**

### CODE REVIEW

File: apps/yara/mapper.h [hunk 1]
```diff
@@ -1089,15 +1089,15 @@ inline void _verifyMatchesImpl(Mapper<TSpec, TConfig> & me, PairedEnd)
...
     typedef typename Iterator<TMatesSet>::Type      TMatesSetIt;
 
     start(me.timer);
     unsigned long anchorsCount = length(me.matchesByCoord);
-    TMatchesAppender appender(me.matchesByCoord);
-    TMatchesVerifier verifier(me.ctx, appender,
+//    TMatchesAppender appender(me.matchesByCoord);
+    TMatchesVerifier verifier(me.ctx, me.matchesByCoord,
                               me.contigs.seqs, me.reads.seqs,
                               me.optimalMatchesSet,
```
Comment (h-2) [inline apps/yara/mapper.h]: remove the commented code in this flle

File: apps/yara/mapper_filter.h [hunk 2]
```diff
@@ -49,7 +49,7 @@ using namespace seqan;
 template <typename TSpec, typename Traits>
 struct FilterDelegate
 {
-    typedef typename Traits::THitsAppender  THits;
+    typedef typename Traits::THits  THits;
```
Comment (h-2) [inline apps/yara/mapper_filter.h]: correct the column layout

File: include/seqan/parallel/parallel_atomic_primitives.h [hunk 3]
```diff
@@ -411,6 +410,7 @@ inline T1 * atomicAdd(T1 * volatile & x, T2 y)
 // Wrappers to use faster non-synced functions in serial implementations
 // ----------------------------------------------------------------------------
 
+// NOTE(rrahn): The wrapper functions are deprecated and should be replaced by stl atomic library.
 template <typename T>   inline T atomicInc(T          & x,             Serial)      { return ++x;                    }
```
Comment (h-2) [inline include/seqan/parallel/parallel_atomic_primitives.h]: also mark as `[[deprecated]]` so the compiler will warn

File: include/seqan/parallel/parallel_queue.h [hunk 4]
```diff
@@ -101,7 +101,8 @@ class ConcurrentQueue
     TAtomicSize tailPos;                    char pad6[SEQAN_CACHE_LINE_SIZE - sizeof(TAtomicSize)];
     TAtomicSize tailWritePos;               char pad7[SEQAN_CACHE_LINE_SIZE - sizeof(TAtomicSize)];
     TAtomicSize roundSize;                  char pad8[SEQAN_CACHE_LINE_SIZE - sizeof(TAtomicSize)];
-    Atomic<bool>::Type virgin;              char pad9[SEQAN_CACHE_LINE_SIZE - sizeof(Atomic < bool > ::Type)];
+    // TODO(rrahn): Make this to an atomic_flag?
```
Comment (h-2) [inline include/seqan/parallel/parallel_queue.h]: ?

File: include/seqan/parallel/parallel_sequence.h [hunk 5]
```diff
@@ -40,147 +41,195 @@
...
 {
-    me.data_host = &string;
+    auto tmp = lhs.data_begin.load();
+    lhs.data_begin.store(rhs.data_begin.load());
+    rhs.data_begin.store(tmp);
+    tmp = lhs.data_end.load();
+    lhs.data_end.store(rhs.data_end.load());
+    rhs.data_end.store(tmp);
+    std::swap(lhs.data_capacity, rhs.data_capacity);
+    // NOTE(rrahn): The lock should not be swapped.
```
Comment (h-2) [inline include/seqan/parallel/parallel_sequence.h]: ?

**Comment by h-2** [comment]
OK. What I think is still missing:
* [ ] remove dead code
* [ ] decide whether to remove things like atomicInc or just mark them as deprecated
* [ ] extensive bam_io and yara test
* [ ] eternal happiness

File: include/seqan/parallel/parallel_sequence.h [hunk 6]
```diff
@@ -40,147 +41,195 @@
...
             {
-                valueConstruct(begin(string, Standard()) + newLen - 1, val);
+                valueConstruct(begin(me, Standard()) + newLen - 1, std::forward<TValue>(val));
                 break;
             }
-            _decLength(string, Parallel());
+            _decLength(me);
         }
-
+
```
Comment (h-2) [inline include/seqan/parallel/parallel_sequence.h]: remove trailing whitespace

**Comment by h-2** [comment]
Also there is a build failure for the parallel test and also dox, although I don't see how that is related.

**Comment by rrahn** [comment]
I know, I fixed these issues already, but there is another problem with the qgram index, which uses atomicCas for standard integer. I can not easily replace this with an atomic. So I have to check this.

**Comment by rrahn** [comment]
Seems like small performance increase for bam_io:

**Comment by rrahn** [comment]
Mhmm, this involves too much changes in some parts of the library, which need to be tested thoroughly. Won't happen tonight 😩, sorry.

**Comment by h-2** [comment]
So we close this?

**Pull request closed without merging**